storage/pruner/pruner.go

```diff
@@ -0,0 +1,374 @@
+		}
+
+		// baseHeight is the highest height for which we have completed all processing
+		baseHeight := p.baseHeight()
+		if baseHeight < p.lowestHeader.Height {
+			// this should never happen and means there's a bug. the lowest fully processed height
+			// cannot be lower than the lowest height in the database
+			ctx.Throw(errors.New("inconsistent state: base height must not be lower than the lowest height"))
+			return
+		}
+
+		lg := p.log.With().Uint64("baseHeight", baseHeight).Logger()
+
+		// if there are not enough blocks to prune, skip
+		if baseHeight-p.lowestHeader.Height < p.pruneThreshold {
+			lg.Debug().
+				Uint64("lowestHeight", p.lowestHeader.Height).
+				Msg("not enough blocks to prune. skipping pruning")
+			continue
+		}
+
+		baseHeader, err := p.headers.ByHeight(baseHeight)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("could not get header for height %d: %w", baseHeight, err))
+			return
+		}
+
+		err = p.prune(ctx, baseHeader)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+			ctx.Throw(err)
+			return
+		}
+
+		// finally, increment the lowest height
+		p.lowestHeader = baseHeader
+
+		lg.Info().Msg("pruning complete")
+	}
+}
+
+// prune removes data for all blocks up to the given header exclusive.
+// No errors are expected during normal operation.
+func (p *Pruner) prune(ctx context.Context, baseHeader *flow.Header) error {
+	batch := pebbleStorage.NewBatch(p.db)
+	defer func() {
+		if err := batch.Close(); err != nil {
+			p.log.Error().Err(err).Msg("failed to close batch")
+		}
+	}()
+
+	if p.pruneMethod == PruneByView {
+		if err := p.pruneToView(ctx, baseHeader.View, batch); err != nil {
+			return fmt.Errorf("could not prune to view: %w", err)
+		}
+	} else {
+		if err := p.pruneToHeight(ctx, baseHeader.Height, batch); err != nil {
+			return fmt.Errorf("could not prune to height: %w", err)
+		}
+	}
+
+	if err := batch.Flush(); err != nil {
+		return fmt.Errorf("could not commit batch: %w", err)
+	}
+
+	return nil
+}
+
+// pruneBlock removes all data associated with a block
+// this method contains all the logic to perform the actual pruning.
+// No errors are expected during normal operation.
+func (p *Pruner) pruneBlock(blockID flow.Identifier, batch storage.BatchStorage) error {
+	time.Sleep(DefaultThrottleDelay)
+
+	err := p.chunkDataPacks.Prune(blockID, batch)
+	if err != nil {
+		return fmt.Errorf("could not prune chunk data packs: %w", err)
+	}
+
+	return nil
+}
+
+// pruneToView removes all data for all blocks up to but excluding the given view
+// forks are ignored since all blocks contained within them will eventually be pruned by view.
+// No errors are expected during normal operation.
+func (p *Pruner) pruneToView(ctx context.Context, lastView uint64, batch storage.BatchStorage) error {
+	currentView := p.lowestHeader.View
+	for currentView < lastView {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		header, err := p.headers.ByView(currentView)
```

Comment:
I think the reason you need both `pruneByView` and `pruneByHeight` is to be able to prune blocks on non-finalized fork, which can only be referred by the view not by height.

pruneByHeight is also nice, because the height, unlike the view, is guaranteed to increment sequentially without gap.

The tricky thing is if a non-finalized block is pruned, do we prune its children cascadingly? I would say no, as that's easy to implement and reason about. 

In that case,  we can divide the logic into two parts: 
1. GetAllBlockIDsAtHeight
2. for each block prune it by block ID

For 1, we could use `headers.ByParentID(header.ParentID)`. We could consider adding this method to the headers module. 

This way we don't need to maintain additional `View -> BlockID` index.

Reply:
@AlexHentschel suggested the prune by view approach. The idea is that it simplifies the logic by avoiding the need to traverse forks. Instead, forks are left in place and will eventually be cleaned up by view. We can then ignore the forks entirely.

The reason I added the prune by height in this PR was to support testing on nodes that don't already have a block/view index.

Reply:
> The tricky thing is if a non-finalized block is pruned, do we prune its children cascadingly? I would say no, as that's easy to implement and reason about.
>
> In that case, we can divide the logic into two parts:
> 
> 1. GetAllBlockIDsAtHeight
> 2. for each block prune it by block ID
> 
> For 1, we could use `headers.ByParentID(header.ParentID)`. We could consider adding this method to the headers module.
> 
> This way we don't need to maintain additional `View -> BlockID` index.

to implement `GetAllBlockIDsAtHeight`, wouldn't we also need to capture all blocks in the forks? Otherwise, as soon as we delete the first ancestor in the chain, we lose the ability to find the children at the next iteration.

Reply:
Yeah, I'm just afraid there are some index based on height. So if we prune by view, then we can't prune things by height. For instance, register store indexes registers by height in storage, not by view nor by blockID. And you can't lookup height by view, because it's not 1:1 mapping.

Reply:
> to implement GetAllBlockIDsAtHeight, wouldn't we also need to capture all blocks in the forks? Otherwise, as soon as we delete the first ancestor in the chain, we lose the ability to find the children at the next iteration.

Yeah, that's a good point. In that case, prune by view is better. However, we need to make sure the view-blockid index exists, which require either a dynamic bootstrap or a migration. So be careful about this. 

We can do a combination of PruneByView and PruneByHeight:
We first find block by the view, and then check if it's a finalized block. If not, then just prune the block; if yes, then also prune other indexes by its height.